### PR TITLE
mimic: krbd: fix rbd map hang due to udev return subsystem unordered

### PR DESCRIPTION
**Pull request by smithfarm**
backport tracker: https://tracker.ceph.com/issues/39316

---

backport of https://github.com/ceph/ceph/pull/27339
parent tracker: https://tracker.ceph.com/issues/39089

this backport was staged using https://github.com/ceph/ceph/blob/master/src/script/ceph-backport.sh

### CODE REVIEW

File: src/krbd.cc [hunk 1]
```diff
@@ -222,37 +224,50 @@ static int wait_for_udev_add(struct udev_monitor *mon, const char *pool,
             this_image && strcmp(this_image, image) == 0 &&
             this_snap && strcmp(this_snap, snap) == 0) {
           bus_dev = dev;
-          continue;
+          goto check;
```
Comment (smithfarm) [inline src/krbd.cc]: This conditional block is where I had to resolve conflicts manually, and I'm not super confident the conflict resolution is correct.

Comment (idryomov) [reply src/krbd.cc]: This looks fine.

File: src/krbd.cc [hunk 2]
```diff
@@ -222,37 +224,50 @@ static int wait_for_udev_add(struct udev_monitor *mon, const char *pool,
...
-        assert(!minor ^ have_minor_attr());
+      for (auto p : block_dev_vec) {
+        const char *this_major = udev_device_get_property_value(p, "MAJOR");
+        const char *this_minor = udev_device_get_property_value(p, "MINOR");
 
         if (strcmp(this_major, major) == 0 &&
             (!minor || strcmp(this_minor, minor) == 0)) {
           string name = get_kernel_rbd_name(udev_device_get_sysname(bus_dev));
 
           assert(strcmp(udev_device_get_devnode(dev), name.c_str()) == 0);
```
Comment (idryomov) [inline src/krbd.cc]: This is wrong: the original patch changes `udev_device_get_devnode(dev)` to `udev_device_get_devnode(p)`.

Comment (smithfarm) [reply src/krbd.cc]: oops! fixing

**Comment by smithfarm** [comment]
@idryomov Please have a second look. Thanks.

**Comment by yuriw** [comment]
[wip-yuri4-testing-2019-10-04-0016-mimic](https://trello.com/c/sWFPFDBT/709-wip-yuri4-testing-2019-10-04-0016-mimic)

**Pull request merged**